### PR TITLE
collate now supports minimal dimension + added more support in samples cacher + added interface validation utility

### PR DESCRIPTION
**Pull request by YoelShoshan**

### CODE REVIEW

File: fuse/data/pipelines/pipeline_default.py [hunk 1]
```diff
@@ -138,7 +138,7 @@ def __call__(
             if self._verbose:
                 context = Timer(
                     f"Pipeline {self._name}: op {type(op).__name__}, op_id {sub_op_id}",
-                    self._verbose,
+                    # self._verbose,
```
Comment (mosheraboh) [inline fuse/data/pipelines/pipeline_default.py:141]: revert

File: fuse/data/datasets/caching/samples_cacher.py [hunk 2]
```diff
@@ -418,6 +451,28 @@ def _cache(self, orig_sample_id: Any) -> Any:
         return output_info
 
 
+def _convert_to_sequence_for_hdf5_if_needed(key: str, data: Any) -> Dict:
+    ans = {}
+    if isinstance(data, (list, tuple)):
+        if isinstance(data, list):
+            ans[key + "@RESERVED_LIST@"] = np.array(
```
Comment (mosheraboh) [inline fuse/data/datasets/caching/samples_cacher.py:458]: Why are you wrapping the length with np.array?

**Review by mosheraboh — DISMISSED** [review]
Looks good.
Few minors inline

File: fuse/data/utils/collates.py [hunk 3]
```diff
@@ -134,7 +134,7 @@ def _batch_dispatch(
             batch_dict[key] = self._special_handlers_keys[key](collected_values)
         elif isinstance(
             collected_values[0],
-            (torch.Tensor, np.ndarray, float, int, str, bytes),  # , tuple),
+            (torch.Tensor, np.ndarray, float, int, str, bytes, tuple),
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply fuse/data/utils/collates.py:137]: removed tuple 👍

File: fuse/data/utils/collates.py [hunk 4]
```diff
@@ -175,6 +182,17 @@ def pad_all_tensors_to_same_size(
         # get max per dim
         max_per_dim = np.amax(np.stack([value.shape for value in values]), axis=0)
 
+        if min_size_per_dim is not None:
+            assert isinstance(min_size_per_dim, tuple)
+            assert len(min_size_per_dim) == len(min_size_per_dim)
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply fuse/data/utils/collates.py:187]: (was supposed to be `assert len(min_size_per_dim) == len(max_per_dim)` )

File: fuse/data/pipelines/pipeline_default.py (hunk 1 above)
Comment (YoelShoshan) [reply fuse/data/pipelines/pipeline_default.py:141]: Timer does not accept any verbose arg

Comment (mosheraboh) [reply fuse/data/pipelines/pipeline_default.py:141]: So delete self._verbose 😆

File: fuse/utils/interface_validator.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                curr_arg
+            ]
+
+        # then replace ourself with the original method
+        setattr(owner, name, curr_func)
+
+
+class InterfaceValidator:
+    def __init__(self, verbose: bool = False):
+        self.__dict__["_properly_initialized_InterfaceValidator"] = True
```
Comment (mosheraboh) [inline fuse/utils/interface_validator.py:86]: super().__init__() is recommded.
Why self.__dict__["_properly_initialized_InterfaceValidator"] and not self._properly_initialized_InterfaceValidator ?

File: fuse/utils/interface_validator.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        for method_name, curr_validate_args in self._validate_methods.items():
+            curr_bases = self.__class__.__bases__
+            if len(curr_bases) > 1:
+                raise Exception("currently not supporting more than one base")
+            if len(curr_bases) == 0:
+                raise Exception("expected at least one base")
+
+            if getattr(type(self), method_name) == getattr(curr_bases[0], method_name):
+                raise Exception(
+                    f'a class that inherited from InsilicoLabModel did not override interface method "{method_name}" !'
```
Comment (mosheraboh) [inline fuse/utils/interface_validator.py:102]: Do you mean InterfaceValidator?

File: fuse/utils/interface_validator.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+```
+
+the API will be enforced and exceptions will be raised if requirements are not met.
+
+
+#TODO:
+    1. allow validating return type hint
+    2. allow using @validate_signature without using any args, and that would mean that the exact same full signature is required
+    3. allow control over whether type hint is validated or not (is this needed?)
+"""
```
Comment (mosheraboh) [inline fuse/utils/interface_validator.py:50]: Also validating that there are no additional arguments without default values is important

File: fuse/utils/interface_validator.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+    def _validate_signature(
+        self, method_name: str, validate_args: Dict[str, _UnionGenericAlias]
+    ) -> None:
+        """
+        raises an exception if a signature mismatch is found
+        """
+        curr_method = getattr(self, method_name)
+        actual_method = curr_method
+        # if hasattr(curr_method, '__wrapped__'): #the standard functools.wraps does not keep type annotations!
```
Comment (mosheraboh) [inline fuse/utils/interface_validator.py:120]: delete

**Review by mosheraboh — APPROVED** [review]
Looks great!
See my comments inline.

File: fuse/data/pipelines/pipeline_default.py (hunk 1 above)
Comment (YoelShoshan) [reply fuse/data/pipelines/pipeline_default.py:142]: done

File: fuse/data/utils/collates.py [hunk 9]
```diff
@@ -153,6 +153,7 @@ def pad_all_tensors_to_same_size(
         values: List[torch.Tensor],
         pad_val: float = 0.0,
         min_size_per_dim: Optional[Tuple] = None,
+        verbose: bool = False,
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply fuse/data/utils/collates.py:156]: deleted

File: fuse/data/utils/collates.py [hunk 10]
```diff
@@ -184,7 +185,7 @@ def pad_all_tensors_to_same_size(
 
         if min_size_per_dim is not None:
             assert isinstance(min_size_per_dim, tuple)
-            assert len(min_size_per_dim) == len(min_size_per_dim)
+            assert len(min_size_per_dim) >= len(max_per_dim)
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply fuse/data/utils/collates.py:188]: seems safer as "==" - changed to it, thanks!

File: fuse/utils/interface_validator.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+    return wrapper
+
+
+class validate_signature_impl(object):
+    def __init__(self, fn: Callable, args: List[str] = None) -> None:
+        self.fn = fn
+        self._args_to_validate = args
+
+    def __set_name__(self, owner: any, name: str) -> None:
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply fuse/utils/interface_validator.py:65]: yes - thanks! fixed

File: fuse/utils/interface_validator.py (hunk 5 above)
Comment (YoelShoshan) [reply fuse/utils/interface_validator.py:86]: if you use `self._properly_initialized_InterfaceValidator` then you will immediately get into `self.__setattr__`  and get an exception that it doesn't exist yet

Comment (YoelShoshan) [reply fuse/utils/interface_validator.py:86]: "super().init() is recommded" - do you mean that I should switch from using `super().__init__()` to `super().init()` ?

File: fuse/utils/interface_validator.py (hunk 6 above)
Comment (YoelShoshan) [reply fuse/utils/interface_validator.py:102]: yes, it's leftover from before I extracted this class, thanks! fixed

File: fuse/utils/interface_validator.py (hunk 7 above)
Comment (YoelShoshan) [reply fuse/utils/interface_validator.py:50]: can you elaborate, I don't think I understand what you meant

File: fuse/utils/interface_validator.py (hunk 5 above)
Comment (YoelShoshan) [reply fuse/utils/interface_validator.py:86]: about super().__init__() ok I think I understand what you meant, that it's just generally a good practice, maybe even when you don't inherit from anything

**Review by mosheraboh — APPROVED** [review]
LGTM

**Pull request merged**